src/cargo/util/graph.rs

```diff
@@ -1,52 +1,219 @@
+            }
+        };
+    }
+
+    /// connect `node`to `child` associating it with `edge`.
+    /// Note that if this and `link` are used on the same graph
+    ///      odd things may happen when `reset_to` is called.
+    pub fn add_edge(&mut self, node: N, child: N, edge: E) {
+        use indexmap::map::Entry;
+        let edge = Some(edge);
+        // IndexMap happens to do exactly what we need to keep the ordering correct.
+        match self
+            .nodes
+            .entry(node.clone())
+            .or_insert_with(IndexMap::new)
+            .entry(child.clone())
+        {
+            Entry::Vacant(entry) => {
+                // add the new edge, and link and fix the new link count
+                self.edges.push(EdgeLink {
+                    value: edge,
+                    next: None,
+                    previous: None,
+                });
+                let edge_index: EdgeIndex = self.edges.len() - 1;
+                entry.insert(edge_index);
+            }
+            Entry::Occupied(entry) => {
+                // this pare is already linked
+                let mut edge_index = *entry.get();
+                loop {
+                    // follow the linked list
+                    if self.edges[edge_index].value == edge {
+                        return;
+                    }
+                    if self.edges[edge_index].next.is_none() {
+                        // we found the end, add the new edge
+                        self.edges.push(EdgeLink {
+                            value: edge,
+                            next: None,
+                            previous: Some(edge_index),
+                        });
+                        let new_index: NonZeroEdgeIndex =
+                            NonZeroUsize::new(self.edges.len() - 1).unwrap();
+                        // make the list point to the new edge
+                        self.edges[edge_index].next = Some(new_index);
+                        return;
+                    }
+                    edge_index = self.edges[edge_index].next.unwrap().get();
+                }
+            }
+        }
     }
 
     pub fn contains<Q: ?Sized>(&self, k: &Q) -> bool
     where
         N: Borrow<Q>,
-        Q: Ord + Eq,
+        Q: Hash + Eq,
     {
         self.nodes.contains_key(k)
     }
 
-    pub fn edge(&self, from: &N, to: &N) -> Option<&E> {
-        self.nodes.get(from)?.get(to)
+    pub fn iter(&self) -> impl Iterator<Item = &N> {
+        self.borrow().nodes.keys()
+    }
+
+    pub fn edge(&self, from: &N, to: &N) -> Edges<'_, E> {
+        Edges {
+            graph: &self.edges,
+            index: self.nodes.get(from).and_then(|x| x.get(to).copied()),
+        }
     }
 
-    pub fn edges(&self, from: &N) -> impl Iterator<Item = &(N, E)> {
-        self.nodes.get(from).into_iter().flat_map(|x| x.iter())
+    pub fn edges(&self, from: &N) -> impl Iterator<Item = (&N, Edges<'_, E>)> {
```

Comment:
This seems somewhat odd, but the idea here is that you can implicitly associate a list of data with an edge between nodes? I guess this was previously handled where `E` was `Vec<T>` and nowadays `E` is just `T`, right?

Reply:
Yes. The old Graph can only have one `E` per link. So `E` was `Rc<Vec<Dependency>>`. Most of the time there is only 1 Dependency; the 2 cache misses did not by us mutch. The new Graph can store multiple `E` per link, so it needs to provide an iterator. The disadvantages of storing them in a linked list only come on the rare occasion that we have more than one Dependency. The other advantage is that `reset_to` is not possible (at zero cost) if we hand out `&mut E`.